index.js

```diff
@@ -0,0 +1,101 @@
+const regExp = require('time-limited-regular-expressions')();
+
+module.exports = {
+  extend: 'apostrophe-forms-base-field-widgets',
+  label: 'Regular Expression Text Input',
+  afterConstruct: function(self) {
+    self.addFieldType();
+    self.pushExtraAssets();
+    self.pushExtraSingleton();
+  },
+  beforeConstruct: function(self, options) {
+    options.addFields = [
+      {
+        name: 'regExp',
+        label: 'Regular Expression',
+        type: 'apostrophe-forms-regexp',
+        htmlHelp: 'This must be a valid <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions">regular expression</a>. Example: ^[a-z]{3}$ matches three lowercase letters exactly.',
+        required: true
+      },
+      {
+        name: 'invalid',
+        label: 'Message Displayed When Invalid',
+        type: 'string',
+        help: 'This should be a short message suitable to display next to the field.',
+        required: true
+      }
+    ].concat(options.addFields || []);
+  },
+  construct: function (self, options) {
+    const settings = (options.arrangeFields || []).find(group => group.name === 'settings');
+    if (settings) {
+      settings.fields.push('regExp', 'invalid');
+    }
+
+    self.sanitizeFormField = async function(req, form, widget, input, output) {
+      const value = self.apos.launder.string(input[widget.fieldName]);
+      if (value.length) {
+        try {
+          if (!(await regExp.match(widget.regExp, value))) {
+            throw {
+              fieldError: {
+                field: widget.fieldName,
+                error: 'invalid',
+                errorMessage: widget.invalid
+              }
+            };
+          }
+        } catch (e) {
+          if (e.name === 'timeout') {
+            throw {
+              fieldError: {
+                field: widget.fieldName,
+                error: 'invalid',
+                errorMessage: 'Your input took too long to parse. Please check whether you followed the instructions.'
```

Comment:
Could this actually be user error? Maybe this should be configurable so a dev could include a message about where to contact them if it's repeatedly timing out.

Reply:
This is the right error message in the unlikely situation where it is genuine user error. I think making it configurable could be a nice to have but isn't a blocker for this module being published.